rotkehlchen/chain/ethereum/eth2.py

```diff
@@ -88,17 +148,108 @@ def get_eth2_staked_amount(
                     validator_index=int.from_bytes(decoded_data[4], byteorder='little'),
                     tx_hash=tx_hash,
                     log_index=event['logIndex'],
+                    timestamp=Timestamp(transaction.timestamp),
                 ))
-                totals[transaction.from_address] += amount
                 break
+    return deposits
+
+
+def get_eth2_staked_amount(
+        ethereum: 'EthereumManager',
+        addresses: List[ChecksumEthAddress],
+        has_premium: bool,
+        msg_aggregator: MessagesAggregator,
+        database: 'DBHandler',
+) -> Eth2DepositResult:
+    """Get the addresses' ETH2 staked amount
+
+    For any given new address an on-chain query from the ETH2 deposit contract
+    deployment timestamp until now will be run.
+
+    For any existing address an on-chain query from the minimum "to timestamp"
+    last used query range among all the existing addresses until now will be run.
+
+    Then write in DB all the new deposits and finally return them all.
+    """
+    new_deposits: List[Eth2Deposit] = []
+    totals: Dict[ChecksumEthAddress, Balance] = {}
+    new_addresses: List[ChecksumEthAddress] = []
+    existing_addresses: List[ChecksumEthAddress] = []
+    to_ts = ts_now()
+    min_from_ts = to_ts
+
+    # Get addresses' last used query range for ETH2 deposits
+    for address in addresses:
+        entry_name = f'{ETH2_DEPOSITS_PREFIX}_{address}'
+        deposits_range = database.get_used_query_range(name=entry_name)
+
+        if not deposits_range:
+            new_addresses.append(address)
+        else:
+            existing_addresses.append(address)
+            min_from_ts = min(min_from_ts, deposits_range[1])
+
+    # Get deposits for new addresses
+    if new_addresses:
+        deposits_ = _get_eth2_staked_amount_onchain(
+            ethereum=ethereum,
+            addresses=new_addresses,
+            has_premium=has_premium,
+            msg_aggregator=msg_aggregator,
+            from_ts=ETH2_DEPLOYED_TS,
+            to_ts=to_ts,
+        )
+        new_deposits.extend(deposits_)
+
+        for address in new_addresses:
+            entry_name = f'{ETH2_DEPOSITS_PREFIX}_{address}'
+            database.update_used_query_range(
+                name=entry_name,
+                start_ts=ETH2_DEPLOYED_TS,
+                end_ts=to_ts,
+            )
+
+    # Get new deposits for existing addresses
+    if existing_addresses and min_from_ts <= to_ts:
```

Comment:
I think I had mentioned this in the standup too, but we **don't** want to rerun these kind of queries again and again even for a difference of 1 second. 

Either use `CACHE_RESPONSE_FOR_SECS` (10 mins for now) or use a new constant for 1 hour. So if saved data is older than an hour ... only then query again.


Do you think that makes sense?